lib/longleaf/helpers/selection_options_parser.rb

```diff
@@ -0,0 +1,162 @@
+          else
+            logger.failure("Invalid checksums parameter format, see `longleaf help <command>` for more information")
+            exit 1
+          end
+        end
+
+        file_paths = options[:file].split(/\s*,\s*/)
+        selector = FileSelector.new(file_paths: file_paths, app_config: app_config_manager)
+      elsif !options[:location].nil?
+        storage_locations = options[:location].split(/\s*,\s*/)
+        selector = FileSelector.new(storage_locations: storage_locations, app_config: app_config_manager)
+        digest_provider = SingleDigestProvider.new(nil)
+      else
+        logger.failure("Must provide one of the following file selection options: -f, l, or -m")
+        exit 1
+      end
+
+      [selector, digest_provider]
+    end
+
+    def self.there_can_only_be_one(failure_msg, options, *names)
+      got_one = false
+      names.each do |name|
+        if !options[name].nil?
+          if got_one
+            logger.failure(failure_msg)
+            exit 1
+          end
+          got_one = true
+        end
+      end
+    end
+
+    # Parses the provided manifest options, reading the contents of the manifests to produce
+    # a mapping from files to one or more algorithms.
+    # @param manifest_vals [Array] List of manifest option values. They may be in one of the following formats:
+    #       <alg_name>:<manifest_path> OR <alg_name>:@-
+    #.      <manifest_path> OR @-
+    # @return a hash containing the aggregated contents of the provided manifests. The keys are
+    #    paths to manifested files. The values are hashes, mapping digest algorithms to digest values.
+    def self.manifests_to_digest_mapping(manifest_vals)
+      alg_manifest_pairs = []
+      # interpret option inputs into a list of algorithms to manifest sources
+      manifest_vals.each do |manifest_val|
+        if manifest_val.include?(':')
+          manifest_parts = manifest_val.split(':', 2)
+          alg_manifest_pairs << manifest_parts
+        else
+          # algorithm no specified in option value
+          alg_manifest_pairs << [nil, manifest_val]
+        end
+      end
+      if alg_manifest_pairs.select { |mpair| mpair[1] == '@-' }.count > 1
+        self.fail("Cannot specify more than one manifest from STDIN")
+      end
+
+      # read the provided manifests to build a mapping from file uri to all supplied digests
+      path_to_digests = Hash.new { |h,k| h[k] = Hash.new }
+      alg_manifest_pairs.each do |mpair|
+        source_stream = nil
+        # Determine if reading from a manifest file or stdin
+        if mpair[1] == '@-'
+          source_stream = $stdin
+        else
+          source_stream = File.new(mpair[1])
+        end
+
+        current_alg = mpair[0]
+        if current_alg.nil?
+          # No algorithm specified, manifest must specify algorithms internally using headers
+          source_stream.each_line do |line|
+            line = line.strip
+            if /[a-zA-Z0-9]+:/ =~ line
```

Comment:
Should we make this more restrictive to reduce the likelihood of matching hashes that happen to contain colons? For example, with `^[a-zA-Z0-9]+:$`? I think this still may allow a hash like `abcxyz123:` to break things, raising an exception in `DigestHelper.validate_algorithms()`?

Reply:
Fixing this might also make it possible to merge the repeating manifest parsing code.

Reply:
Refactored into a single case/loop. Also added `DigestHelper.is_known_algorithm?` for checking it is a valid algorithm rather than throwing an uncaught error if it isn't.